User can view an item

tpetersen0308 opened the pull request:
# Things Done
- Create Item Page for each item
- Extract countdown utility and add tests 
- Rename folder structure and components name to be more clear

# Live Render
<img width="922" alt="Screen Shot 2019-07-19 at 9 37 26 AM" src="https://user-images.githubusercontent.com/29721784/61550992-d914f380-aa08-11e9-8dcd-086aa30d1d3d.png">

tpetersen0308 committed:
Display individual item info when the user visits an item show page.

Render individual item.

serpient commented on src/components/item/Container.js:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  fetchItem = async () => {
+    const self = this;
```
I'm wondering why you are saving a reference to `this`?
doesn't the big arrow share the reference to `this` ?

serpient commented on src/components/item/Container.js:
```diff
@@ -0,0 +1,51 @@
+  render() {
+    return (
+      !this.state.loading &&
+      <Item props={this.state} />
```
At this current moment, it doesn't seem like we need state / class component. Since its a simple fetch and pass of data into a <Item /> component, maybe we can use a functional component + React Hooks ?

serpient commented on src/components/item/Container.test.js:
```diff
@@ -0,0 +1,34 @@
+  component.getByText('test description');
+  component.getByText('2019-07-17T16:53:52Z');
+  component.getByText('Current Bid: $10');
+  component.getByText('test title');
```
I'm a little unfamiliar with this test, but what does this assert? Just that the particular text exists in the component?

serpient commented on src/components/item/Presenter.js:
```diff
@@ -0,0 +1,25 @@
+      <img data-testid={`item-image-${props.id}`} src={props.image} alt={props.title} />
+      <div className={`item-info`}>
+        <div className="item-title">
+          <span id={`item-title-${props.title}`}><strong>{props.title}</strong></span>
```
Once there are multiple usages of a key (for example, `props.title`), it's probably a good time to start object restructuring to make it easier to change later

aaizuss replied:
To expand on what Francesca says, you can add `const { title, price, endDate } = props` before your return, and then you won't have to reference `props.price`, `props.title`, etc in multiple places

tpetersen0308 committed:
Extract date countdown function to utility module

tpetersen0308 committed:
Rename files and folders to be more intuitive

serpient committed:
Update styling

- Will need to revise bidding area later for final styling touches and add media queries

serpient committed:
Add links from item card to item page

- Add Header link, when clicked, to redirect back to "/"
- Adjust styling for links
- Add responsive styling to Item Page

tpetersen0308 committed:
Add error handling to Item

zacholauson commented on src/utilities/date.js:
```diff
@@ -0,0 +1,11 @@
+  hourDifference = hourDifference < 0 ? 24 + hourDifference : hourDifference;
+
+  if (dayDifference < 0) {
+    return "This Auction Has Ended";
```
This presentation messaging is pretty specific, while the function seems like it should be pretty generic.

tpetersen0308 replied:
We'll break these up into separate functions

zacholauson commented on src/components/item_page/ItemPage.scss:
```diff
@@ -0,0 +1,59 @@
+@import 'styles/_shared.scss';
```
I don't have a preference, but just out of curiosity, is it more idiomatic to use styled React components vs external CSS?

jayden replied:
I think it's more about personal preference than what's considered idiomatic, but I personally like using styled components.

zacholauson commented on src/components/item_page/ItemPage.js:
```diff
@@ -0,0 +1,30 @@
+          </div>
+          <div className="item-bidding-info">
+            <span id={`item-price-${price}`}>Current Bid: ${price}</span><br />
+            <span id={`item-end-date-${end_date}`}>{countdown(new Date(), new Date(end_date))}</span><br />
```
`countdown` seems to always take "now" as the first argument, is that something worth pushing into the function? Or maybe in the form of another function like `countdownUntil(futureDate)`?

tpetersen0308 replied:
We initially did that, but passing in the start date allowed us to test more directly. Is that a bad reason to make that design choice?

zacholauson replied:
Nah, driving your design using tests is great. In this case, I think having a wrapping function that can delegate would be a decent approach here.

jayden commented:
Was `package-lock.json` accidentally deleted? It should stay in source control so that the package versions are locked.

jayden commented on src/components/item_card/ItemCard.js:
```diff
@@ -0,0 +1,20 @@
+import { countdown } from 'utilities/date';
+
+const ItemCard = ({ item }) => {
+    let { title, price, image, end_date, id } = item;
```
Is there a reason why we'd want to use `let` here instead of `const`? This comment also applies to every other usage of `let`.

jayden commented on src/components/item_page/Item.js:
```diff
@@ -0,0 +1,28 @@
+
+  useEffect(() => {
+    const fetchItems = async () => {
+      let result = await Fetch(`/api/items/${props.match.params.id}`);
```
We can collapse these two lines into one:
```
const { data, error } = await Fetch(`/api/items/${props.match.params.id}`);
```

jayden commented on src/components/item_page/Item.js:
```diff
@@ -0,0 +1,28 @@
+
+
+  const handleError = (err) => {
+    return err && "Error: " + err.message;
```
Can we use string interpolation here for consistency?

zacholauson commented on src/components/item_page/Item.test.js:
```diff
@@ -0,0 +1,57 @@
+
+describe('ItemsPage Test', () => {
+  beforeEach(() => {
+    jest.mock('axios');
```
I see that you are wrapping `axios` in `utilities/fetch`, any reason you are mocking `axios` directly here?
https://8thlight.com/blog/eric-smith/2011/10/27/thats-not-yours.html
https://github.com/testdouble/contributing-tests/wiki/Don't-mock-what-you-don't-own

tpetersen0308 replied:
We'll address this in PR #4

zacholauson commented:
> Was `package-lock.json` accidentally deleted? It should stay in source control so that the package versions are locked.

It looks like it's still here, just with lots of lines removed so the diff got truncated.

tpetersen0308 committed:
Refactor date utility.

tpetersen0308 committed:
Refactor item components.

- Change let to const where possible.
- Use object destructuring to handle fetch results.
- Use string interpolation in handleError().

tpetersen0308 committed:
Resolve merge conflicts.
